text/3803-enum-repr-no-niches.md

```diff
@@ -0,0 +1,709 @@
+- Feature Name: `enum_repr_no_niches`
+- Start Date: 2025-04-21
+- RFC PR: [rust-lang/rfcs#3803](https://github.com/rust-lang/rfcs/pull/3803)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+
+# Summary
+[summary]: #summary
+
+Allow `#[repr(_, no_niches)]` on field-less (C-style) enums to disable the niche layout optimization on discriminants, allowing using the enum directly in FFI.
```

Comment:
I think “no niche optimization” is a poor framing here, because in general “disable optimization” isn’t a sturdy of foundation on which to build anything, and also because talking to me about niches doesn’t tell me about the exhaustiveness change unless I think about it hard. I think it would be better to:

* Call the repr modifier something like “`allow_undeclared_variants`”.
* Reframe the RFC text around “this makes it allowed (not UB) to have a value of an undeclared variant” instead of about niches.

Reply:
I especially think `no_niches` is the wrong framing for this, because the main use case is for FFI, where niches aren't used anyway, since neither passing the enum directly nor using it in a repr(c) struct allow use of any niches. 

I think the salient point here not that the compiler can't use niches for the type, but that the behavior for discrimant values outside of the defined variants is still well defined, that is it doesn't trigger Undefined Behavior.

Reply:
> How about the clang's terminology: `#[open]`

I proposed `#[repr(open)]` elsewhere on this PR, but thinking of it a bit more, I think `#[open]` might be a more appropriate option because it affects the semantics of the enum in a deeper way than, say, `#[repr(C)]` or `#[repr(u8)]`.